_posts/2021-09-16-drain-cleaner.md

```diff
@@ -0,0 +1,196 @@
+continue uninterrupted during a Kubernetes upgrade, where each node is drained right after the previous one has been 
+fully evacuated.
+
+The problem is that node draining does not account for the actual internal state of the Kafka cluster, as it works using
+the `Readiness` state, which indicates if the container is ready to start accepting traffic, but does not necessarily
+correspond to the replicas being in sync. For that reason, there is no guarantee that the next node draining will start
+before the previous evacuated Kafka Pod is fully synced, so you might end up with under-replicated partitions for some
+time, causing the producer's rate to drop to zero. This can affect the entire data pipeline, so also consumers will 
+likely be affected.
+
+## Leveraging admission webhooks
+
+The [DrainCleaner](https://github.com/strimzi/drain-cleaner) is a lightweight application that leverages admission 
+webhooks to help move Kafka Pods from Kubernetes nodes which are being drained. This is useful if you want the 
+ClusterOperator to move Pods instead of Kubernetes itself. The advantage of this approach is that the ClusterOperator 
+makes sure that no partition becomes under-replicated during node draining, by waiting for the synchronization of topics 
+with multiple replicas and the value of `min.insync.replicas` less than `replicas`.
+
+As you can see from the following image, when a new Kubernetes request arrives at the API server endpoint, a 
+[ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers) 
+controller is invoked after the schema validation phase and it calls every registered webhook, to implement some 
+logic. This can be useful to enforce custom policies or simply be notified about specific events.
+
+![admission controllers](/assets/images/posts/2021-09-16-admission-controllers.png)
+
+Some admission controllers are actually built-in (i.e. LimitRange, NamespaceLifecycle). ValidatingWebhooks are safe, as 
+they cannot change the incoming request and they always see the final version that would be persisted to etcd. 
+
+> Note: ValidatingWebhooks may be disabled in your Kubernetes cluster for security reasons, so this is the first thing to verify.
+
+In our case, we want the DrainCleaner to be notified of all Pod eviction requests. The application will then filter out 
+all Pods that are not part of our StatefulSet. One important note is that the `Kafka` custom resource must have 
+`maxUnavailable: 0` in Kafka and ZooKeeper [PodDisruptionBudgets](https://kubernetes.io/docs/concepts/workloads/Pods/disruptions/). 
+This is required to block Pod draining and let the combined work of DrainCleaner and the ClusterOperator handle this.
+
+This is how you can apply that configuration with a single patch command (no rolling update is needed):
+
+```sh
+kubectl patch kafka $CLUSTER_NAME --type json -p '[
```

Comment:
You should use something like `<CLUSTER-NAME>` instead of `$CLUSTER_NAME` to make it more clear this needs to be replaced.